pkg/controller/registry/resolver/sat/solve.go

```diff
@@ -0,0 +1,95 @@
+package sat
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/irifrance/gini"
+)
+
+// Identifier values uniquely identify particular Installables within
+// the input to a single call to Solve.
+type Identifier string
+
+// Installable values are the basic unit of problems and solutions
+// understood by this package.
+type Installable interface {
+	// Identifier returns the Identifier that uniquely identifies
+	// this Installable among all other Installables in a given
+	// problem.
+	Identifier() Identifier
+	// Constraints returns the set of constraints that apply to
+	// this Installable.
+	Constraints() []Constraint
+}
+
+var Incomplete = errors.New("cancelled before a solution could be found")
+
+// NotSatisfiable is an error composed of a minimal set of applied
+// constraints that is sufficient to make a solution impossible.
+type NotSatisfiable []AppliedConstraint
+
+func (e NotSatisfiable) Error() string {
+	const msg = "constraints not satisfiable"
+	if len(e) == 0 {
+		return msg
+	}
+	s := make([]string, len(e))
+	for i, a := range e {
+		s[i] = a.String()
+	}
+	return fmt.Sprintf("%s: %s", msg, strings.Join(s, ", "))
+}
+
+const (
+	satisfiable   = 1
+	unsatisfiable = -1
+	timeout       = 0
```

Comment:
nit: is this used?

Reply:
Not anymore. OK with you to mop that up in a future PR?